File: app/common/python/projection/datastore/projection_datastore_proxy.py
from typing import Dict, List

from app.common.python.projection.datastore.projection_datastore_shard import ProjectionDatastoreShard

MAX_SHARD_COUNT = 100


class ProjectionDatastoreProxy:
    '''
    This class is used as a proxy to the database that stores our
    projection.
    '''

    def __init__(self, shard_size: int = 25):
        self.shards = []
        self.movie_indices = dict()
        self.shard_size = shard_size
        self.shard_count = len(self.shards)

    def _create_new_shard(self):
        new_shard = ProjectionDatastoreShard()
        self.shards.append(new_shard)
        self.shard_count += 1

    def _cache_data(self, movie_indices: Dict[str, int]) -> None:
        self.movie_indices = movie_indices

    def _upload_to_shards(self, projection: Dict[str, List[float]]) -> None:
        ordered_projection = list(projection.items())
        projection_size = len(projection)
        offset = 0
        shard_index = 0
        while offset < projection_size:
            # Create a new shard if we need one
            if shard_index is self.shard_count:
                self._create_new_shard()

            projection_for_shard = dict()
            batch_size = min(self.shard_size, projection_size - offset)
            for batch_index in range(batch_size):
                author, vector = ordered_projection[offset + batch_index]
                projection_for_shard[author] = vector
            self.shards[shard_index].upload(projection_for_shard)
            shard_index += 1
            offset += batch_size

    def upload(self, projection: Dict[str, List[float]], movie_indices: Dict[str, int]) -> None:
        self._cache_data(movie_indices)
        self._upload_to_shards(projection)

    def get_movie_indices(self) -> Dict[str, int]:
        return self.movie_indices

    def get_shards(self) -> List[ProjectionDatastoreShard]:
        return self.shards

File: experimental/steven/cpp/datastores/projection_datastore_proxy.cpp
#include "projection_datastore_proxy.h"

std::unique_ptr<ProjectionDatastoreProxy> ProjectionDatastoreProxy::create(
    ProjectionFileStore file_store, std::string projection_filepath_root,
    std::string movie_indices_filepath) {
  auto datastore = ProjectionDatastoreProxy(
      file_store, projection_filepath_root, movie_indices_filepath, SHARD_SIZE);
  datastore.loadData();
  return std::make_unique<ProjectionDatastoreProxy>(datastore);
}

ProjectionDatastoreProxy::ProjectionDatastoreProxy(
    ProjectionFileStore file_store, std::string projection_filepath_root,
    std::string movie_indices_filepath, int shard_size)
    : file_store(file_store),
      projection_filepath_root(projection_filepath_root),
      movie_indices_filepath(movie_indices_filepath),
      shard_size(shard_size),
      shard_count(0) {}

std::vector<ProjectionDatastoreShard>& ProjectionDatastoreProxy::getShards() {
  return this->shards;
}

std::map<std::string, int>& ProjectionDatastoreProxy::getMovieIndices() {
  return this->movie_indices;
}

void ProjectionDatastoreProxy::upload(
    std::map<std::string, std::vector<float>> projection,
    std::map<std::string, int> movie_indices) {
  this->saveData(movie_indices);
  this->cacheData(movie_indices);
  this->uploadToShards(projection);
}

void ProjectionDatastoreProxy::cacheData(
    std::map<std::string, int> movie_indices) {
  this->movie_indices = movie_indices;
}

std::string ProjectionDatastoreProxy::getShardFilepath(int shard_index) {
  return this->projection_filepath_root + std::to_string(shard_index) +
         std::string(".json");
}

void ProjectionDatastoreProxy::createNewShard() {
  int id = this->createShardId();
  std::string filepath = this->getShardFilepath(id);
  auto new_shard = ProjectionDatastoreShard(file_store, filepath);
  this->shards.push_back(new_shard);
}

int ProjectionDatastoreProxy::createShardId() { return shard_count++; }

// TODO this can be multithreaded, createNewShard uses an atomic counter to be
// thread safe.
void ProjectionDatastoreProxy::initializeShards() {
  this->shard_count = 0;
  this->shards = std::vector<ProjectionDatastoreShard>();
  while (this->shard_count < MAX_SHARDS &&
         this->file_store.objectExists(
             this->getShardFilepath(this->shard_count))) {
    this->createNewShard();
  }
}

void ProjectionDatastoreProxy::loadMovieIndices() {
  if (this->file_store.objectExists(this->movie_indices_filepath)) {
    this->movie_indices =
        this->file_store.getMovieIndices(this->movie_indices_filepath);
  }
}

void ProjectionDatastoreProxy::saveData(
    std::map<std::string, int> movie_indices) {
  this->file_store.putMovieIndices(this->movie_indices_filepath, movie_indices);
}

void ProjectionDatastoreProxy::loadData() {
  this->loadMovieIndices();
  this->initializeShards();
  for (auto& shard : this->shards) {
    shard.loadData();
  }
}

void ProjectionDatastoreProxy::uploadToShards(
    std::map<std::string, std::vector<float>> projection) {
  int projection_size = projection.size();
  int offset = 0;
  std::atomic<int> shard_index = 0;
  // TODO this can be done in parallel.
  while (offset < projection_size) {
    // Create a new shard if we need one
    if (shard_index == this->shard_count) {
      this->createNewShard();
    }

    std::map<std::string, std::vector<float>> projection_for_shard;
    int batch_size = std::min(this->shard_size, projection_size - offset);
    for (int batch_index = 0; batch_index < batch_size; batch_index++) {
      std::string author;
      std::vector<float> ratings;
      projection_for_shard[author] = ratings;
    }
    this->shards[shard_index++].upload(projection_for_shard);
    offset += batch_size;
  }
}